### DISCUSSION TITLE
403 forbidden error when trying to send form action on production version (built) of SvelteKit app

### DISCUSSION THREAD

**Discussion by michaelusantiago** [pocketbase/pocketbase#1301] [Q&A]
My app is running fine on development mode, but when I build the app `npm run build` and run it `node ./app/index.js` which will run the app on `0.0.0.0:3000` then the app will run but when I try to log_in (send form actions), it throw errors in web console the `Failed to load resource: the server responded with a status of 403 (Forbidden).` PocketBase is running on `127.0.0.1:8090`. I honestly not sure if this is PocketBase issue, but I'm trying. Hoping to get help.

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand the issue.

Could you provide the exact error network response? You can find it in your browser's "Inspect tool > Network" tab.

Alternatively you can also start PocketBase with `--debug` and/or check the logs in the "Admin UI > Logs" (you could filter the requests using something like `status != 200`).

**Reply by michaelusantiago** [reply]
@ganigeorgiev I even thought the pocketbase not even reach when the app is running in production mode. My issue is this, I have a SveleteKit App that I build `npm run build` ready for deployment, the development version is working fine, but when I run the production version or build version this error is appearing in the web console.

![forbidden 403](https://user-images.githubusercontent.com/4829195/208380973-9ab64a7b-9737-46bf-8b1d-e5ccf14a1b68.png)

I even try to run the pocketbase `pocketbase serve --http="0.0.0.0:8090" ` to match the frontend host.

I will try the --debug

**Reply by michaelusantiago** [reply]
![forbidden 403 2](https://user-images.githubusercontent.com/4829195/208381886-3ac2d525-671a-4922-a645-a264da035254.png)

**Reply by michaelusantiago** [reply]
There's no logs in the PocketBase UI about it because, I said, I thought the pocketbase is not even reach.

**Reply by ganigeorgiev** [reply]
The above screenshot shows that you are sending a request to your node server, not directly to PocketBase, so my guess is that the 403 is returned from node.

Could you provide a code sample how you are sending and processing the request and how the JS SDK is initialized?

**Reply by michaelusantiago** [reply]
Wait, I'm sorry, I actually getting a logs from PocketBase UI but the result is 200 and 204. So I am only getting this in the frontend.

**Reply by michaelusantiago** [reply]
![forbidden 403 3](https://user-images.githubusercontent.com/4829195/208383219-24e960b7-4ecc-4291-a400-a57b278d5eac.png)

**Reply by michaelusantiago** [reply]
Here's I send the log_in form action.

```ts
export const actions: Actions = {
    // LOG IN
    log_in: async ({ request, locals }) => {
        const formData = await request.formData();
        const email = formData.get("email") as string;
        const password = formData.get('password') as string;

        try {
            /** Will authenticate user and will add cookie when successful
             *  or throw error if not */
            await locals.pb.collection('users').authWithPassword(email, password);

            return ({ success: true });
        } catch (error: any) {
            console.log(error)
            return ({
                err: true,
                message: serializeNonPOJ(error?.data)
            });
        }
    },
. . . .
```

**Reply by ganigeorgiev** [reply]
@michaelusantiago As far as I understand, the issue is `?/log_in` call. The logs in the console seems to be from the realtime endpoint, not from the `auth-with-password`.

**Reply by michaelusantiago** [reply]
I don't get it, where did I gone wrong there. The app/site is running well when in development mode.

I'm not even directly callling ?/log_in than sending a Form action POST. (SvelteKit).

**Reply by ganigeorgiev** [reply]
@michaelusantiago Could you confirm that the above `log_in` action is even triggered (eg. add a print statement after the `formData` parsing)? From the code snippet, if this was a PocketBase issue it should have triggered your `catch` handling.

**Reply by michaelusantiago** [reply]
this is the from that trigger the actions log_in

```ts
<form method="POST" use:enhance = {
({ form, data, action }) => {
    loading = true;
    return async ({ result, update }) => {
        loading = false;
        log_in_result = result;
        if (log_in_result?.data?.success) {
            $show_login_form = false;
            location.reload();
        } else log_in_error = log_in_result?.data;
    }
}}>
<h1 class="font-bold text-2xl">Log In</h1>
<label>
    Email
    <input
        name="email"
        type="email"
        bind:this={email}
        on:change={() => log_in_error = null}
    />
</label>
<label>
    Password
    <input
.  .  .
```

I'm not calling or using fetch, just the pocketbase JS SDK as written in the documentation.

**Reply by michaelusantiago** [reply]
> @michaelusantiago Could you confirm that the above `log_in` action is even triggered (eg. add a print statement after the `formData` parsing)? From the code snippet, if this was a PocketBase issue it should have triggered your `catch` handling.

It is, it works in development version of the app. Ok I will show console.log in there, but in development mode.

**Reply by ganigeorgiev** [reply]
> I'm not calling or using fetch, just the pocketbase JS SDK as written in the documentation.

It is most likely sveltekit that is using fetch to call your form action. My suggestion was to check whether you are even triggering the `log_in` action or the 403 is triggered somewhere before reaching the endpoint. As mentioned previously, if the error was from PocketBase your `catch` handler will take for it (and you'll have log in the Admin UI).

> It is, it works in development version of the app. Ok I will show console.log in there, but in development mode.

Without access to the code, I'm not sure how to help debugging it. If you are still not able to resolve the issue, if possible, please try to create a minimal reproducible example and I'll try to check it in more details.

**Reply by michaelusantiago** [reply]
![log email](https://user-images.githubusercontent.com/4829195/208386366-09ec7d0f-6a54-40cc-86d8-4fe302f1a4f8.png)

I now believe it reach PocketBase base on the screenshot of PocketBase UI I showed, the referrer is indeed coming from localhost:3000 which is the production version of the App.

**Reply by ganigeorgiev** [reply]
> now believe it reach PocketBase base on the screenshot of PocketBase UI 

Are you sure that you are sending the correct email and password?
You can find a more detailed error message in the Admin UI > Logs and/or when starting PocketBase with `--debug`.

_Also please note that the above error is 400 not 403._

**Reply by michaelusantiago** [reply]
I have no problem with that, I can log in correctly on the app if in the development mode. I only show in the last screenshot that I can log inside the `log_in` from action in `page.server.ts` of my app.

**Reply by michaelusantiago** [reply]
The code I show you is when running the app in development mode. That's the only way I can show and display logs. No logs are showing when in production version of the app.

**Reply by michaelusantiago** [reply]
My code/source code of the app is in my github, can you check it out?

**Reply by ganigeorgiev** [reply]
Could you share the link?

**Reply by michaelusantiago** [reply]
Sure, please, I put it in public, It's my portfolio website actually.

https://github.com/michaelusantiago/michaelsantiago.ca

**Reply by ganigeorgiev** [reply]
@michaelusantiago I'm not able to reproduce issue (node 18.10). What I did:
 
- git clone
- npm install
- npm run dev
- adjusted my collections to have a similar structure based on the fields required in the `layout.server.ts`
- npm run build
- npm run preview
-> the log_in action is triggered in my case

I'm not sure how to help. There are way too many things in the code that I don't understand and it is hard to navigate and pinpoint the exact reason for the error. My advice would be to try creating a minimal reproducible example that shows only the error with the minimum collections configuration that are required for your code to run.

**Reply by michaelusantiago** [reply]
@ganigeorgiev npm run preview also works for me. I;m getting issue only when this is the way `node ./app/inde.js`

Have your tried to run the build version, not eh `npm run preview`?

**Reply by michaelusantiago** [reply]
Try the `node ./app/index.js` not the `npm run preview`. The build output is in `app` folder.

**Reply by ganigeorgiev** [reply]
In this case I believe the issue is related to https://github.com/sveltejs/kit/tree/master/packages/adapter-node#environment-variables.

Once again, check the exact error message in the browser "Inspect tool > Network" tab. In the response you probably have something like: "POST form submissions are forbidden".

If that's the case, then you can try starting node with explicit `ORIGIN` value, eg.:

```
ORIGIN=http://localhost:3000 node ./app/index.js
```

**Reply by ganigeorgiev** [reply]
Per the above documentation:

> HTTP doesn't give SvelteKit a reliable way to know the URL that is currently being requested. The simplest way to tell SvelteKit where the app is being served is to set the ORIGIN environment variable

**Reply by michaelusantiago** [reply]
I run in it `pocketbase serve --debug`

![log debug pb](https://user-images.githubusercontent.com/4829195/208434388-6bfc3714-175c-45a2-bee3-a7564cf25c7f.png)

I tried to login from the form and get the same error. I can confirm now that it reach PocketBase everytime I send a Form POST/actions. Just that the browsers throw this 403 forbidden error. I honestly don't know where is this coming from. I ask in discord in Sveltekit group and somebody respond that the backend I am using is sending this 403, though I have a feeling that it is SvelteKit issue.

**Reply by ganigeorgiev** [reply]
@michaelusantiago Did you try starting node with `ORIGIN=http://localhost:3000`? 

If yes and you are still experiencing the 403 error, please provide the exact error message from the browser "Inspect tool > Network > Response" tab.

Also, if you haven't already, make sure to check https://github.com/sveltejs/kit/tree/master/packages/adapter-node#environment-variables.

**Reply by michaelusantiago** [reply]
I did that too, I will try it again.

**Reply by michaelusantiago** [reply]
I have read the links you send already, and still anylyzing.

**Reply by ganigeorgiev** [reply]
@michaelusantiago What is the exact 403 error message response?

**Reply by michaelusantiago** [reply]
One thing though is I am not using .env. I don't have environment variables used in this website. So case of Environment Variable is eliminated already.

**Reply by ganigeorgiev** [reply]
But have you tried setting the `ORIGIN=http://localhost:3000` before starting node, aka:

```
ORIGIN=http://localhost:3000 node ./app/index.js
```
?

**Reply by ganigeorgiev** [reply]
But once again, the best way to debug it is to check the exact 403 error message response.

**Reply by michaelusantiago** [reply]
![log origin](https://user-images.githubusercontent.com/4829195/208437133-5b9cc45a-fbab-45f2-8b23-6ca137f33a00.png)

this is not working command

**Reply by ganigeorgiev** [reply]
@michaelusantiago This is a windows/cmdlet issue. I'm not familiar with the terminal you are using. You have to find a way to set an env variable (either during the build process or before run).

But to confirm that the issue is caused by the missing `ORIGIN`, please check the 403 error response in the browser Inspect tool > Network tab > Response

**Reply by michaelusantiago** [reply]
got, it I run it in a gitbash instead. Its my quick reply. I will send more details... Just want to let you know I run the command on gitbash successfully.

**Reply by michaelusantiago** [reply]
I didnt' get errors on the browser now. NO more 403 and I received a console.log in my terminal. more details...

**Reply by michaelusantiago** [reply]
You're the man! :) Appreciated you help bro! It is working now, atleast.. :)

The key is `ORIGIN=https://localhost:3000 node ./app/index.js` on a `bash` terminal. Looking to see if it will behave when I dockerized things in a docker compose.

**Reply by ganigeorgiev** [reply]
I'm glad that you were able to resolve the issue, but please check carefully again the adatapter-node documentation https://github.com/sveltejs/kit/tree/master/packages/adapter-node, because there are other node specifics like the `BODY_SIZE_LIMIT` (which defaults to 512kb) that may need to be adjusted. If you provide a .env file during build (aka. `node -r dotenv/config build`), setting manually the env variables before run may not be necessary.

**Reply by michaelusantiago** [reply]
> I'm glad that you were able to resolve the issue, but please check carefully again the adatapter-node documentation https://github.com/sveltejs/kit/tree/master/packages/adapter-node, because there are other node specifics like the `BODY_SIZE_LIMIT` (which defaults to 512kb) that may need to be adjusted. If you provide a .env file during build (aka. `node -r dotenv/config build`), setting manually the env variables before run may not be necessary.

Thanks man, ! will, I noticed that also, I will read this documentation, this I haven't saw before building my app.

Anyway, man, I have another inquiry like I can't limit a record to query, The last time I tried, it's not working. I hope we have discord like group where we can interactively answer inquiries.

`// you can also fetch all records at once via getFullList
const records = await pb.collection('posts').getFullList(200 /* batch size */, {
    sort: '-created',
});`
 I thought this one limits a record to 200 but the last time I tried this it's not. If I change the batch size to 3, it still get me the full list. I know it's not connected to the issue I raised, but this is the quickest I thought I can get.

It's a great experience though to use PocketBase. Keep up the good work man! Looking forward to contribute someday.

**Reply by ganigeorgiev** [reply]
`getFullList` returns all items. `200` is the batch requests size, aka. how many items per requests to fetch until all items are downloaded.

If you want a one-off paginated list, use `pb.collection('posts').getList(1, 3, { sort: "-created" })`.

**Reply by michaelusantiago** [reply]
> `getFullList` returns all items. `200` is the batch requests size, aka. how many items per requests to fetch until all items are downloaded.
> 
> If you want a one-off paginated list, use `pb.collection('posts').getList(1, 3, { sort: "-created" })`.

Yeah that's what I used/did the paginated one instead. Oh so, I misunderstood the `getfullList()`. I see, great!

**Comment by iamshcc** [comment]
You can fix it by disabling the CSRF origin check in your `svelte.config.js` file:

```javascript
import adapter from '@sveltejs/adapter-node';

/** @type {import('@sveltejs/kit').Config} */
const config = {
  kit: {
    adapter: adapter(),
    // disable CSRF origin check (local dev only, don’t do this in prod)
    csrf: {
      checkOrigin: false
    }
  }
};

export default config;
```

Setting checkOrigin: false tells SvelteKit to skip CSRF origin validation.
This is fine for local or private setups, but for public deployments you should set up proper CORS and CSRF protection.
